src/main/java/ch/njol/skript/sections/EffSecSpawn.java

```diff
@@ -0,0 +1,161 @@
+
+import java.util.List;
+
+// TODO this won't show up in the docs, sections don't have a tab. We should create a tab for them,
+//  and maybe add EffectSections to the effects page as well
+@Name("Spawn")
+@Description({"Spawn a creature. This can be used as an effect and as a section.",
+	"If it is used as a section, the section is run before the entity is added to the world.",
+	"You can modify the entity in this section, using for example 'event-entity' or 'cow'. ",
+	"Do note that other event values, such as 'player', won't work in this section."})
+@Examples({"spawn 3 creepers at the targeted block",
+	"spawn a ghast 5 meters above the player",
+	"spawn a zombie at the player:",
+	"\tset name of the zombie to \"\""})
+@Since("1.0, INSERT VERSION (with section)")
+public class EffSecSpawn extends EffectSection {
+
+	public static class SpawnEvent extends Event {
+		private final Entity entity;
+
+		public SpawnEvent(Entity entity) {
+			this.entity = entity;
+		}
+
+		public Entity getEntity() {
+			return entity;
+		}
+
+		@Override
+		public @NotNull HandlerList getHandlers() {
+			throw new IllegalStateException();
+		}
+	}
+
+	static {
+		Skript.registerSection(EffSecSpawn.class,
+			"(spawn|summon) %entitytypes% [%directions% %locations%]",
+			"(spawn|summon) %number% of %entitytypes% [%directions% %locations%]");
+		EventValues.registerEventValue(SpawnEvent.class, Entity.class, new Getter<Entity, SpawnEvent>() {
+			@Override
+			public Entity get(SpawnEvent spawnEvent) {
+				return spawnEvent.getEntity();
+			}
+		}, 0);
+	}
+
+	@Nullable
+	public static Entity lastSpawned = null;
+
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<Location> locations;
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<EntityType> types;
+	@Nullable
+	private Expression<Number> amount;
+
+	@Nullable
+	private Trigger trigger;
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public boolean init(Expression<?>[] exprs,
+						int matchedPattern,
+						Kleenean isDelayed,
+						ParseResult parseResult,
+						@Nullable SectionNode sectionNode,
+						@Nullable List<TriggerItem> triggerItems) {
+		amount = matchedPattern == 0 ? null : (Expression<Number>) (exprs[0]);
+		types = (Expression<EntityType>) exprs[matchedPattern];
+		locations = Direction.combine((Expression<? extends Direction>) exprs[1 + matchedPattern], (Expression<? extends Location>) exprs[2 + matchedPattern]);
+
+		if (sectionNode != null)
+			trigger = loadCode(sectionNode, "spawn", SpawnEvent.class);
+
+		return true;
+	}
+
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	@Override
+	protected @Nullable TriggerItem walk(Event e) {
+		lastSpawned = null;
+
+		Consumer<? extends Entity> consumer;
+		if (trigger != null) {
+			consumer = o -> {
+				SpawnEvent spawnEvent = new SpawnEvent(o);
+				// Copy the local variables from the calling code to this section
+				Variables.setLocalVariables(spawnEvent, Variables.copyLocalVariables(e));
```

Comment:
Maybe we should copy local variables outside this consumer, since it's possible the consumer may be ran several times.